include/ringmesh/mesh/volume_mesh.h

```diff
@@ -0,0 +1,304 @@
+         * @return the cell facet center
+         */
+        vecn< DIMENSION > cell_facet_barycenter(
+            const CellLocalFacet& cell_local_facet ) const;
+
+        /*!
+         * Compute the non weighted barycenter of the \param cell_id
+         */
+        vecn< DIMENSION > cell_barycenter( index_t cell_id ) const;
+
+        /*!
+         * Computes the Mesh cell facet normal
+         * @param[in] cell_local_facet the cell index and
+         * the local facet index in the cell
+         * @return the cell facet normal
+         */
+        vecn< DIMENSION > cell_facet_normal(
+            const CellLocalFacet& cell_local_facet ) const;
+
+        /*!
+         * @brief compute the volume of the cell \param cell_id.
+         */
+        virtual double cell_volume( index_t cell_id ) const = 0;
+
+        std::vector< index_t > cells_around_vertex(
+            index_t vertex_id, index_t cell_hint ) const;
+
+        index_t find_cell_corner( index_t cell_id, index_t vertex_id ) const;
+
+        bool find_cell_from_colocated_vertex_within_distance_if_any(
+            const vecn< DIMENSION >& vertex_vec,
+            double distance,
+            index_t& cell_id,
+            index_t& cell_vertex_id ) const;
+
+        /*!
+         * @brief return the NNSearch at cell facets
+         * @warning the NNSearch is destroyed when calling the
+         * Mesh::facets_aabb()
+         *  and Mesh::cells_aabb()
+         */
+        const NNSearch< DIMENSION >& cell_facet_nn_search() const;
+
+        /*!
+         * @brief return the NNSearch at cells
+         */
+        const NNSearch< DIMENSION >& cell_nn_search() const
+        {
+            if( !cell_nn_search_ )
+            {
+                std::vector< vecn< DIMENSION > > cell_centers( nb_cells() );
+                for( auto c : range( nb_cells() ) )
+                {
+                    cell_centers[c] = cell_barycenter( c );
+                }
+                cell_nn_search_.reset(
+                    new NNSearch< DIMENSION >( cell_centers, true ) );
+            }
+            return *cell_nn_search_.get();
+        }
+        /*!
+         * @brief Creates an AABB tree for a Mesh cells
+         */
+        const VolumeAABBTree< DIMENSION >& cell_aabb() const
+        {
+            if( !cell_aabb_ )
+            {
+                cell_aabb_.reset( new VolumeAABBTree< DIMENSION >( *this ) );
+            }
+            return *cell_aabb_.get();
+        }
+
+        bool is_mesh_valid() const override;
+        std::tuple< index_t, std::vector< index_t > >
+            connected_components() const final;
+
+    protected:
+        VolumeMesh() = default;
+
+    protected:
+        mutable std::unique_ptr< NNSearch< DIMENSION > >
+            cell_facet_nn_search_{};
+        mutable std::unique_ptr< NNSearch< DIMENSION > > cell_nn_search_{};
+        mutable std::unique_ptr< VolumeAABBTree< DIMENSION > > cell_aabb_{};

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)